src/libraries/System.Security.Cryptography/tests/OpenSslNamedKeysTests.manual.cs

```diff
@@ -174,17 +225,202 @@ public static void Engine_UsePublicKey()
+            using ECDsa ecdsaPri = new ECDsaOpenSsl(priKeyHandle);
+
+            byte[] data = new byte[] { 1, 2, 3, 1, 1, 2, 3 };
+
+            byte[] ecdsaPubBytes = ecdsaPri.ExportSubjectPublicKeyInfo();
+            ECDsa ecdsaPub = ECDsa.Create();
+            ecdsaPub.ImportSubjectPublicKeyInfo(ecdsaPubBytes, out int bytesRead);
+            Assert.Equal(ecdsaPubBytes.Length, bytesRead);
+
+            using ECDsa ecdsaBad = ECDsa.Create();
+            ecdsaBad.KeySize = ecdsaPri.KeySize;
+
+            // Verify can sign/verify multiple times
+            for (int i = 0; i < 10; i++)
+            {
+                data[0] = (byte)i;
+                byte[] signature = ecdsaPri.SignData(data, HashAlgorithmName.SHA256);
+                byte[] badSignature = ecdsaBad.SignData(data, HashAlgorithmName.SHA256);
+                Assert.NotEqual(data, signature);
+                Assert.NotEqual(data, badSignature);
+                Assert.NotEqual(badSignature, signature);
+                Assert.True(ecdsaPub.VerifyData(data, signature, HashAlgorithmName.SHA256));
+                Assert.False(ecdsaPub.VerifyData(data, badSignature, HashAlgorithmName.SHA256));
+                Assert.False(ecdsaBad.VerifyData(data, signature, HashAlgorithmName.SHA256));
+
+                // TPM key is intended for sign/decrypt only, we could theoretically make verify work without needing to export/import by forcing 'default' provider
+                // for this operation but it's most likely misusage on user part and tpm2 provider intentionally didn't allow it so we will follow this logic.
+                Assert.ThrowsAny<CryptographicException>(() => ecdsaPri.VerifyData(data, signature, HashAlgorithmName.SHA256));
+            }
+
+            // It's TPM so it should not be possible to export parameters
+            Assert.ThrowsAny<CryptographicException>(() => ecdsaPri.ExportParameters(includePrivateParameters: true));
+        }
+
+        [ConditionalFact(nameof(ShouldRunProviderEcDhTests))]
+        public static void Provider_TPM2ECDH()
+        {
+            using SafeEvpPKeyHandle priKeyHandle = SafeEvpPKeyHandle.OpenKeyFromProvider(Tpm2ProviderName, TpmEcDhKeyHandleUri);
+            using ECDiffieHellman alicePri = new ECDiffieHellmanOpenSsl(priKeyHandle);
+            using ECDiffieHellman alicePub = ECDiffieHellman.Create();
+            alicePub.ImportParameters(alicePri.ExportParameters(includePrivateParameters: false));
+
+            using ECDiffieHellman bobPri = ECDiffieHellman.Create();
+            bobPri.KeySize = alicePri.KeySize;
+
+            using ECDiffieHellman bobPub = ECDiffieHellman.Create();
+            bobPub.ImportParameters(bobPri.ExportParameters(includePrivateParameters: false));
+
+            byte[] sharedKeyFromAlice;
+            using (ECDiffieHellmanPublicKey bobPublic = bobPub.PublicKey)
+            {
+                sharedKeyFromAlice = alicePri.DeriveKeyMaterial(bobPublic);
```

Comment:
`DeriveKeyMaterial` is a bad function to use, since it's provider-dependent (thanks ECDiffieHellmanCng with a derived-type property to control how it worked).

You probably want DeriveRawSecretAgreement.